### ISSUE TITLE
Very high load during scrubs, rendering whole system unresponsive.

### ISSUE THREAD

**Issue by ghost** [zfsonlinux/zfs#4528]
Hi there,

I'm running a Gentoo server with _sys-fs/zfs 0.6.5.4-r2_. I have two raidz2 pools in the system, with 10 disks each, one pool with 4TB drives and one with 6TB drives. They are both filled to about 50%. I scrub these pools once per month. During the scrub the load is about 20 on average, often reaching 24.

When the load is as high, also all 8 cores of my Intel(R) Xeon(R) CPU E3-1230 v3 @ 3.30GHz have 100% usage, almost exclusively by **txg_sync**, **z_rd_iss** or **z_rd_int** tasks/threads. The machine is mainly used as a media server and access times become so high that it's becoming unusable. When I want to check on it via ssh it takes a few seconds after I press "arrow up" until the last command from my zsh history comes up. Grepping /var/log/messages to check for warnings or errors takes a few seconds as well.

The whole scrub takes 30 to 40 hours, which is about 120MB/s and seems pretty slow to me. Shouldn't I get read rates far above 300MB/s with a raidz2 and WD Red hard drives?

I don't believe that the scrub should impact my whole system that much. The processes I mentioned above show a niceness of -20 or -19 in htop, so I don't see why they would be able to make the system so overloaded. Right now one of the scrubs is taking place and should take at least another 24 hours. If anyone could tell me what more information to provide in order to get an understanding of what is going wrong here, I would greatly appreciate it.

**Comment by behlendorf** [comment]
@matrixagent if you haven't already I'd suggest setting `zfs recordsize=1M` for a raidz configuration.  Larger blocks should improve the scrub speed but the large block size will only be applied to new files so you won't see any immediate speed up.

**Comment by matrixagent** [comment]
@behlendorf Thanks for the reply. I had not yet set that option. I have done so now and the scrub finished. But regardless of the duration of the scrub, is it really normal that it uses 100% of all available CPU cores?

**Comment by behlendorf** [comment]
@matrixagent the scrub will attempt to keep the disks busy if they're otherwise idle.  Depending on you system spec that might be a large fraction of the available CPU.  You could throttle back the scrub by setting the module option, `zfs_vdev_scrub_max_active` which controls how many scrub IOs are in flight.  By default it's only 2 but you could decrease it to 1.

Under development are  patches to further reduce the CPU usage on some x86_64 system by relying on some CPU specific instructions (AVX2, etc).

**Comment by darkpixel** [comment]
I've noticed something similar on a Proxmox box.  I'm running two VMs.  When the scrub starts, everything runs fine for about 5-10 minutes, then the scrub speed drops way down, the VMs become extremely sluggish, and the CPU load on the box jumps from ~2 to ~16.  Stopping the scrub results in the load dropping back down to ~2 within 60-90 seconds, and the VMs become responsive within ~10 seconds of stopping the scrub.

Unfortunately it happened at a Bad Time and I had to get the box back in service right away, so I wasn't able to do much in the way of troubleshooting.

The setup is a RAIDZ2 pool with 7*2 TB WD2000FYYZ drives and a SanDisk SD6SB1M064G1022I log drive. The CPU is a single socket 4-core Intel(R) Xeon(R) CPU E5-2609 v2 @ 2.50GHz with 64 GB RAM.  The Kernel is 4.2.6-1-pve.

The box needs to be in-service 24/7, but I can schedule down-time for testing.
